### ISSUE TITLE
Board Review: Pipeline policy for OASIS Repeatability headers

### ISSUE THREAD

**Issue by mikekistler** [Azure/azure-sdk#3853]
The Azure API Guidelines say that all operations should be idempotent. This is mainly an issue for POST operations.  For situations where a POST needs special support to be made idempotent, the guidelines recommend  

☑️ YOU SHOULD support repeatable requests according as defined in [OASIS Repeatable Requests Version 1.0](https://docs.oasis-open.org/odata/repeatable-requests/v1.0/repeatable-requests-v1.0.html).

Some teams have begun following this advice and implementing support for these headers in their service.

This raises the question of how the Azure SDKs should support OASIS Repeatability headers. It seems desirable to allow these headers to be created and injected by a pipeline policy rather than ask the caller to pass in appropriate values.  The headers involved are `Repeatability-Request-ID`, which the pipeline policy would set with a freshly minted UUID, and `Repeatability-First-Sent`, which would be set to the current date/time.  The policy would be inserted before the retry policy so the header values remain in-tact for any retries.

**Comment by mikekistler** [comment]
@JeffreyRichter convinced me that this probably does not belong as a pipeline policy and is just a feature that should be added in each of our generators.  So closing this issue.

**Comment by mikekistler** [comment]
Reopening this issue as there is inconsistency in support for repeatability headers across our generators, so I think we need to proscribe the desired support in the general SDK guidelines.

**Comment by mikekistler** [comment]
Proposal:

When repeatability headers are defined for an operation, here's how they should be exposed:

- Do not create method parameters for the request headers. We generally do not want customers to have to think about these headers.

- However, all languages should allow the user to pass "custom headers" and the library should respect repeatability headers passed in through this mechanism. 

- When not passed by the user, the client library should generate and pass values for the Repeatability-Request-ID (a uuid) and Repeatability-First-Sent (IMF fixdate), keeping and resending these values on any retries generated by the pipeline.

- All languages should allow the user to access all headers of the response, so no additional mechanism is needed to access the Repeatability-Result response header.

- Each language should document (somewhere) an example of setting the repeatability request headers and accessing the Repeatability-Result response header.

**Comment by KrzysztofCwalina** [comment]
> However, all languages should allow the user to pass "custom headers" and the library should respect repeatability headers passed in through this mechanism.

Only protocol methods in .NET allow that, and we would like to keep it as such. 

This whole feature (repeatability) looks to me like mainly an implementation detail of the request generation and the retry policy. i.e. it's a generator feature.

**Comment by mikekistler** [comment]
The point of the guidelines is establish a consistent design for all clients within a language and across languages. _How_ it is implemented is a secondary issue, but I agree that having the generator do the work is the preferred option.

I checked the guidelines and it appears that there is no requirement to accept custom headers so if .NET only supports this for protocol methods that seems fine.

**Comment by lmolkova** [comment]
> However, all languages should allow the user to pass "custom headers" and the library should respect repeatability headers passed in through this mechanism.

> All languages should allow the user to access all headers of the response, so no additional mechanism is needed to access the Repeatability-Result response header.


We can have more than one request and response for complex methods such as download in chunks. We should either say that  the guideline applies to one-method-one-request-with-retries case or limit it to the HttpPolicy level. I.e. user can inject HttpPolicy that would have access to request and response headers right away.

Another thing that would be great to define is order of repeatability and auth and redirect policies and whether repeatability headers should be added for auth requests and redirected locations.

**Comment by JeffreyRichter** [comment]
The guideline is for a method that makes 1 (and just 1) service operation. I propose a simplication to these guidelines:

- Any client method that directly invokes a single service operation, must add the the Repeatability-Request-ID (a uuid) and Repeatability-First-Sent (IMF fixdate) request headers before sending the HTTP request down the pipeline. These header values remain the same cross all retries.
  - NOTE: The code to do this would typically be emitted by a code generator based on the presence of these headers in the API contract (Swagger/TypeSpec).

-  Client methods would not offer explicit parameters allowing the caller to explicitly set these headers values.
   - NOTE: If the SDK allows customer code an "escapte hatch" mechanism to set arbitrary headers, then any cutomer-set values for these 2 headers should be used and must not be overwritten by the client method code.

- The SDK's response model must expose the Repeatability-Result response header
   - NOTE: The code generator would typically emit this too as the response header would also be present in the API contract.

**Comment by JeffreyRichter** [comment]
No, you can delete it. I was just addressing KC's comment explicitly but it's fine to remove

-- Jeffrey
________________________________
